PERF: optimize algos.take for repeated calls

jorisvandenbossche opened the pull request:
This PR optimizes our internal `take` algorithm, with the following changes:

- Cache the expensive parts that are independent of the actual array values but only dependent on the dtypes / dimension / axis / fill_value (which is often the same for repeated calls), i.e. `maybe_promote` and `_get_take_nd_function`
- The above is done in the generic `take_nd` function as well, but in addition I also added a `take_1d_array` specialized version that assumes the input is already an array and only deals with 1D. This gives a small further speed-up.

An example use case where the effect of this optimization can be noticed nicely is `unstack` in case of the ArrayManager. This is typically a case where we do multiple `take` calls for 1 column (to split it into multiple columns) and thus call it many times for the same dtype/fill_value.

Using the example from the `reshape.py::SimpleReshape` ASV benchmark (this is an example of homogeneous dtype with a single block, so which is fast with the BlockManager using the `_can_fast_transpose` fastpath in `_unstack_frame`):

```python
arrays = [np.arange(100).repeat(100), np.roll(np.tile(np.arange(100), 100), 25)]
index = pd.MultiIndex.from_arrays(arrays)
df = pd.DataFrame(np.random.randn(10000, 4), index=index)
df_am = df._as_manager("array")
```

Master:
```
In [2]: %timeit df.unstack()
1.79 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit df_am.unstack()
15.1 ms ± 1.3 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

PR:

```
In [2]: %timeit df.unstack()
1.81 ms ± 160 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [3]: %timeit df_am.unstack()
3.44 ms ± 69.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

So for the BlockManager this doesn't matter (it does only a single call to `take_nd` for the column values), but for the ArrayManager version this gives a nice speedup and brings it within 2x of the BlockManager version (for this benchmark).

jorisvandenbossche committed:
PERF: optimize algos.take for repeated calls

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1538,52 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+        func = _take_1d_dict.get(tup, None)
+    elif ndim == 2:
+        if axis == 0:
+            func = _take_2d_axis0_dict.get(tup, None)
```
The content of this function is simply copied out from `_get_take_nd_function` below, no other changes to the code have been made

jbrockmendel replied:
just checking, this is _cut_ from the existing function, not _copied_?  i.e. this isnt introducing a big chunk of duplicate code?

jorisvandenbossche replied:
Yes, cut (I just meant to say that no logic changed inside it)

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -1661,6 +1677,40 @@ def take(arr, indices, axis: int = 0, allow_fill: bool = False, fill_value=None)
+        else:
+            # check for promotion based on types only (do this first because
+            # it's faster than computing a mask)
+            dtype, fill_value = maybe_promote_cached(arr.dtype, fill_value)
```
This function is also copied verbatim out of `take_nd` (to be reused), the only actual code change is to use here `maybe_promote_cached` instead of `maybe_promote`

jorisvandenbossche commented:
As an example benchmark of `unstack` that doesn't use the 1-block fastpath (from `reshape.py::Unstack::time_full_product[category]`):

```
import string

m = 100
n = 50

levels = np.arange(m)
index = pd.MultiIndex.from_product([levels] * 2)
columns = np.arange(n)
indices = np.random.randint(0, 52, size=(m * m, n))
values = np.take(list(string.ascii_letters), indices)
values = [pd.Categorical(v) for v in values.T]
df = pd.DataFrame(values, index, columns)
df_am = df._as_manager("array")
```

Master:

```
In [2]: %timeit df.unstack()
374 ms ± 7.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit df_am.unstack()
306 ms ± 10 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

PR:

```
In [2]: %timeit df.unstack()
309 ms ± 21 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit df_am.unstack()
177 ms ± 27.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

So here it improves for BlockManager as well (and in this case ArrayManager is faster, because it has less overhead creating all the blocks for each column)

(specifically for this case of Categorical data, there is a lot of room for improvement by making the categorical constructor faster for such cases where less validation is needed, but that's another topic)

jorisvandenbossche committed:
fix nd check + fix cache differentiation of int / bool

jorisvandenbossche committed:
fix non-scalar fill_value case

jorisvandenbossche committed:
fix mypy

jorisvandenbossche committed:
try fix mypy

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -1776,6 +1820,43 @@ def take_nd(
+    if fill_value is lib.no_default:
+        fill_value = na_value_for_dtype(arr.dtype, compat=False)
+
+    if isinstance(arr, ABCExtensionArray):
```
arr is annotated as ndarray

jorisvandenbossche replied:
Ah, good catch, the annotation is wrong (initially I was going to make a version that only dealt with ndarrays, because this check is expensive as well. But as long as the ArrayManager stores both, this check needs to be done somewhere). Also need to update the docstring for that.

jbrockmendel replied:
small perf boost if you check for `not isinstance(arr, np.ndarray)` i think

jorisvandenbossche replied:
Ah, yes, good idea. 
(being able to do ``isinstance(arr, ExtensionArray)`` would also already give a speedup compared to the ABC version (while still being more explicit), but that's difficult to import here I suppose)

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1535,73 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+    return func
+
+
+@functools.lru_cache(maxsize=128)
```
any idea how much of the bump comes from caching this vs caching __get_take_nd_function_cached?

jorisvandenbossche replied:
Both are significant. From a simple python profile run before the changes in this PR: `maybe_promote` takes ca 40% and `_get_take_nd_function` ca 25% of the time of `take_nd`

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1535,73 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+    return maybe_promote(dtype, fill_value)
+
+
+def _maybe_promote(dtype, fill_value):
```
might make sense to just do this on maybe_promote directly?

jorisvandenbossche replied:
Yeah, I thought about that as well. I would first need to check where `maybe_promote` is used elsewhere to know if it would benefit performance

(eg if it would be used in a situation where you can typically have a varying fill_value, it might decrease the benefit of caching overall)

jbrockmendel replied:
makes sense.  im optimistic that some nice simplifications in maybe_promote are on the horizon, so we can take a look at caching there if/when those come to fruition

jorisvandenbossche committed:
fix annotation

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1535,73 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+@functools.lru_cache(maxsize=128)
+def __get_take_nd_function_cached(ndim, arr_dtype, out_dtype, axis):
+    """
+    Part of _get_take_nd_function below that doesn't need the mask
```
this coment does't make any sense w/o the PR context. can you put / move a doc-string here. typing a +1

jorisvandenbossche replied:
The "and thus can be cached" on the next line is the essential continuation of the sentence. 
The mask can be an array, and thus is not hashable and thus cannot be used as argument for a cached function.

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1535,73 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+
+    return None
+
+
```
why is the caching not on this function? having too many levels of indirection is -1

jorisvandenbossche replied:
I will clarify the comment above, the `mask_info` argument to this function is not hashable

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -1776,6 +1820,43 @@ def take_nd(
 take_1d = take_nd
 
 
+def take_1d_array(
```
Did you read the docstring? It's a `take_1d` variant specialized for arrays, so I think the name makes some sense. But happy to hear alternatives.

The current `take_1d` is just an alias of `take_nd`, while here I want to have an actual specialized version.

I could also rename all occurrences of `take_1d` to `take_nd` (since they are aliases anyway, that shouldn't matter), so the "take_1d" name is free to use.

jreback replied:
I read the doc-string and I understand the code, but I think this additional is very confusing. You are reading it in the context of this PR, where its not terrible. However looking at the actual code is it is so non-obvious what is going on and appears to be duplicative. Please try to refactor.

jorisvandenbossche replied:
> Looking at the actual code is it is so non-obvious what is going on and appears to be duplicative.

And then the docstring explains why it is duplicative... Do you have a concrete suggestion?

(I already suggested getting rid of the current `take_1d` alias if that helps removing the confusion)

jorisvandenbossche committed:
improve docstrings

jreback commented:
@jorisvandenbossche a general comment. target PRs are much better for this type of refactoring. Sure occasionally bigger ones are needed. But please try to make small incremental changes (I know you probably feel this is a small incremental change), but its not. Meaning break things into multple PRs where you first move things, then change them up. This may seem like more work on your part, and migth be a bit more, but it will make reviews faster and less back and forth and arguing.

jorisvandenbossche commented:
I think indeed this is a rather targetted change. A large part of the diff is caused by copying part of a function in helper functions to be able to reuse it (and I commented on those parts to indicate where this happened). 

Sure I can do the move before this PR.

jbrockmendel commented:
@jorisvandenbossche the timings in the OP look like this makes the non-ArrayManager code very slightly slower (well within margin of error) for this particular use case.  are there other use cases where this improves perf?

jreback commented:
yea pls remove the alias would help

jorisvandenbossche commented:
-> precursor just moving code: https://github.com/pandas-dev/pandas/pull/39728

jorisvandenbossche commented:
> the timings in the OP look like this makes the non-ArrayManager code very slightly slower (well within margin of error) for this particular use case. are there other use cases where this improves perf?

It's not slower, but just small variation in timing for something that didn't change noticeably. At least, the precision of the timing is not good enough to know if it's tiny bit slower or tiny bit faster. I could have repeated it to have the difference the other way around ;))

I didn't check where this would improve performance for the BlockManager noticeably. In principle any case where the performance is dominated by the overhead of `take` (so you need a case where `take` is called many times), but don't know from the top of my head what such a case could be. 

I suppose that you could measure the benefit in a single `take` (eg reindex) call as well.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -1534,40 +1535,79 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+    and dtypes.
+    """
+    func = None
     if ndim <= 2:
```
when does this not hold?

jorisvandenbossche replied:
No idea, this might left-over from the Panel world

jbrockmendel commented:
can you add a comment along the lines of `# TODO(AM): this can be removed if we dont end up using ArrayManager`

jorisvandenbossche commented:
> can you add a comment along the lines of # TODO(AM): this can be removed if we dont end up using ArrayManager

Where do you want me to add that comment? Specifically on the specialized `take_1d_array`? 
I think the hashing is probably useful in general as well.

jorisvandenbossche committed:
faster EA check

jbrockmendel commented:
> Where do you want me to add that comment?

on any of the new functions where itd be accurate

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jorisvandenbossche committed:
rename take_1d_array to take_1d

jorisvandenbossche committed:
add comment about being useful for array manager

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1791,6 +1831,47 @@ def take_nd(
+    """
+    Specialized version for 1D arrays. Differences compared to take_nd:
+
+    - Assumes input (arr, indexer) has already been converted to numpy array / EA
```
can you add these assertions

jreback commented on pandas/core/algorithms.py:
```diff
@@ -1537,40 +1538,79 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+
+
+def _maybe_promote(dtype, fill_value):
+    try:
```
this is very strange to do. can you simply change all of the call of `_maybe_promote` to `_maybe_promote_cached` that need it (e.g. in the code and not the tests).

putting a try/except *inside* here is reversing the paradigm and not good.

jorisvandenbossche replied:
That is not simply possible. We don't know in advance if the fill_value is going to be hashable or not. So that's the reason the fallback is needed.

jreback replied:
sure but the try/except needs to be in the cached method, NOT here. IOW you are now exposing 2 api's, we need to have exactly one.

jorisvandenbossche replied:
But try/except is because of non-hashable fill_values, which thus cannot be *inside* the cached method, that's the whole reason I added the try/except in the first place. 

I am not exposing two different APIs. These are internal helper methods, and `_maybe_promote_cached` is only used for `_maybe_promote`, and it is `_maybe_promote` that is then used internally in the `take` implementation. 
I can add a comment to `_maybe_promote_cached` that this is only the cached part of `_maybe_promote` and not to be used elsewhere.

jreback replied:
can we just simply ban non-hashables from `_maybe_promote`?

jbrockmendel commented:
The sticking point seems to be the complexity introduced by the caching.  two thoughts that might trim that down

1) the hashability rule in maybe_promote is not-quite right.  ATM `maybe_promote(not_object, not_hashable)` raises, but in principle we could have it return `(object, not_hashable)`

2) Since the hashability check can go at the top of maybe_promote, we can then refactor _the rest_ of maybe_promote into  the cached _maybe_promote.

jreback commented:
> The sticking point seems to be the complexity introduced by the caching. two thoughts that might trim that down
> 
> 1. the hashability rule in maybe_promote is not-quite right.  ATM `maybe_promote(not_object, not_hashable)` raises, but in principle we could have it return `(object, not_hashable)`
> 2. Since the hashability check can go at the top of maybe_promote, we can then refactor _the rest_ of maybe_promote into  the cached _maybe_promote.

yes I think 2 is the way to go here. I am all for caching, but we have to be very careful to make sure the underlying function that is cached is the simplest possible.

jreback commented:
can you merge master

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jorisvandenbossche committed:
use take_nd for now

jorisvandenbossche commented:
Updated (and also removed the `take_1d` specialization for now, to focus on the caching, since that both gives the most significant improvement as generates the most discussion). 

> 2\. Since the hashability check can go at the top of maybe_promote, we can then refactor _the rest_ of maybe_promote into the cached _maybe_promote.

OK, so that version looks more or less like this (@jbrockmendel did I understand that correctly?)

```python
def maybe_promote_with_cache_version1(dtype: np.dtype, fill_value=np.nan):
    if not is_hashable(fill_value):
        if not is_object_dtype(dtype):
            raise ValueError("fill_value must be a scalar")
        return dtype, fill_value
    return _maybe_promote_cached(dtype, fill_value, type(fill_value))

@functools.lru_cache(maxsize=128)
def _maybe_promote_cached(dtype, fill_value, fill_value_type):
    ... current implementation of maybe_promote ...
```

while the version I have in this PR basically looks like:

```python
def maybe_promote_with_cache_version2(dtype, fill_value):
    try:
        return _maybe_promote_cached(dtype, fill_value, type(fill_value))
    except TypeError:
        return _maybe_promote(dtype, fill_value) 

@functools.lru_cache(maxsize=128)
def _maybe_promote_cached(dtype, fill_value, fill_value_type):
    return _maybe_promote(dtype, fill_value)

def _maybe_promote(dtype, fill_value):
    ... current implementation of maybe_promote ...
```

(and to be clear the `maybe_promote_with_cache_version1/2` would then become the actual `maybe_promote` that is used elsewhere in pandas, just using those names to distinguish)

Timing those options:

```
In [2]: dtype = np.dtype(float)
   ...: fill_value = np.nan

In [3]: %timeit maybe_promote(dtype, fill_value)   # <--- the current, non-cached version
2.36 µs ± 10 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit maybe_promote_with_cache_version1(dtype, fill_value)
303 ns ± 3.59 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit maybe_promote_with_cache_version2(dtype, fill_value)
201 ns ± 3.91 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

So the version with try/except is clearly faster as first version with caching. But of course both are in the nanoseconds, so in absolute numbers the difference is not that big. 

For the example that I use in the top post, `maybe_promote` gets called 400 times, which in this case means 400 x 100ns = 40µs gain by using the second version, while the full unstack operation takes around 3ms. Which of course is only a small precentage.

So if there is a strong preference for the slightly slower version 1 compared to version 2, I am fine with either.

jreback commented:
@jorisvandenbossche ok to go with your current version, i would just add a comment explaining the try/except (e.g. that non-hashables will except, maybe you did this alreadY).

jbrockmendel commented:
ill take another look this afternoon

jbrockmendel commented:
AFAICT there isnt any functional difference between maybe_promote_with_cache_version1 vs maybe_promote_with_cache_version2, so lets go with the faster one.  can you move it to dtypes.cast and we'll get this done

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jorisvandenbossche committed:
move caching of maybe_promote to cast.py

jorisvandenbossche committed:
move type comment

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jorisvandenbossche committed:
typo

jreback reviewed (approved):
looks fine, can you merge master to make sure this is green.

jbrockmendel commented on pandas/core/array_algos/take.py:
```diff
@@ -177,41 +178,60 @@ def take_2d_multi(
 
 
+@functools.lru_cache(maxsize=128)
+def __get_take_nd_function_cached(ndim, arr_dtype, out_dtype, axis):
```
does this need to be dunder

jorisvandenbossche replied:
Nope, changed

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jorisvandenbossche committed:
ensure deprecation warning is always raised

jorisvandenbossche committed:
single underscore

jorisvandenbossche commented on pandas/core/dtypes/cast.py:
```diff
@@ -568,6 +569,35 @@ def maybe_promote(dtype: np.dtype, fill_value=np.nan):
     ValueError
         If fill_value is a non-scalar and dtype is not object.
     """
+    # TODO(2.0): need to directly use the non-cached version as long as we
+    # possibly raise a deprecation warning for datetime dtype
+    if dtype.kind == "M":
+        return _maybe_promote(dtype, fill_value)
```
This is a bit unfortunate, but to ensure the warning is always shown, we can't use the cached version for datetime data.

I check what would be the fastest option. The most specific check would be `if isinstance(fill_value, date) and not isinstance(fill_value, datetime)`, but `if dtype.kind == "M"` is a bit faster. 
So the trade-off was between faster for all non-M8 dtypes vs faster for M8 (by being able to use the cached version in most cases) but a bit slower for all other dtypes. So I went with the first (fastest for numeric dtypes).

jbrockmendel replied:
how big the is the perf tradeoff?

since stacklevels are a constant hassle, one option would be to take the find_stacklevel function and change it so that instead of hard-coding "astype" it just looks for the first call that isn't from inside (non-test) pandas

jorisvandenbossche replied:
It's not the stacklevel as such, it's the warning itself. With caching, it occurs only once, while otherwise this warning is raised every time you use it. 

The other option would be to check for this case / raise the warning a level higher up (so eg the line we are commenting up), so that other cases still use the cached version.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-perf-take

jreback commented:
thanks @jorisvandenbossche very nice

jbrockmendel commented:
@jorisvandenbossche im looking at pandas/tests/indexing/multiindex/test_indexing_slow.py::test_multiindex_get_loc and finding that disabling the lru_cache cuts the runtime in about half (for BM, about 40% for AM).  Can you confirm?

jbrockmendel commented:
Disabling the cache speeds up the test suite by about 20% for me.

jorisvandenbossche commented:
Do you have a specific reproducer as a small snippet (outside of pytest)? When running those tests, I do indeed see a speedup when disabling the cache, but running that test includes both pytest overhead as several ways we end up calling take (sort_values, the actual MI getitem, in the assert functions, ..), so might be easier with a narrowed down use case to profile it / understand what is going on.

jbrockmendel commented:
Trimming down test_multiindex_get_loc to something minimal-ish:

```
import numpy as np
import pandas as pd

m = 50
n = 1000
cols = ["jim", "joe", "jolie", "joline", "jolia"]

vals = [
    np.random.randint(0, 10, n),
    np.random.choice(list("abcdefghij"), n),
    np.random.choice(pd.date_range("20141009", periods=10).tolist(), n),
    np.random.choice(list("ZYXWVUTSRQ"), n),
    np.random.randn(n),
]
vals = list(map(tuple, zip(*vals)))

df = pd.DataFrame(vals, columns=cols)
mi = df.set_index(cols[:-1])

key = (8, 'h', pd.Timestamp('2014-10-17 00:00:00'), 'S')
i = 0
k = key[i]

mask = df.iloc[:, i] == k

%timeit right = df[mask]
1.57 ms ± 53.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  # <- main
174 µs ± 4.26 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)  # <- cache disabled
```

Removing the date_range from `vals` (and an entry from `cols`) brings the two back to having the cache case outperforming by a few percent.

First guess is that for some reason dt64NaT is causing a cache miss, but calling _maybe_promote_cached with (np.dtype("M8[ns]"), np.datetime64("NaT", "ns), np.datetime64) isn't showing a miss.

jorisvandenbossche commented:
So it seems that it is "taking" the datetimelike values then, so further trimming it down to just that:

```
n = 1000
vals = np.random.choice(pd.date_range("20141009", periods=10).tolist(), n)
ser = pd.Series(vals)
mask = np.random.randint(0, 10, n) == 0

%timeit ser[mask]

df = pd.DataFrame({"col": vals})

%timeit df[mask]
```

Some observations:

- The slowdown only occurs in the dataframe case, not for series. But that's simply because the series case's code path doesn't pass by those cache functions
- It's due to `_maybe_promote_cached`, not the `_get_take_nd_function_cached`
- I checked the time of hashing both `dtype = np.dtype('<M8[ns]')` and `fill_value = np.datetime64('NaT')`, but nothing seems concerning about this
- Also timing explicitly the `_maybe_promote_(cached)` doesn't seem to show anything:

```
In [1]: dtype = np.dtype('<M8[ns]')
   ...: fill_value = np.datetime64('NaT')

# version without cache
In [2]: %timeit pd.core.dtypes.cast._maybe_promote(dtype, fill_value)
2.8 µs ± 49.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [3]: %timeit pd.core.dtypes.cast.maybe_promote(dtype, fill_value)
468 ns ± 7.62 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

jbrockmendel commented:
Thanks for taking a look.  I get similar numbers when profiling just the _?maybe_promote functions.

Speculation: each time we call maybe_promote with (np.dtype("M8[ns]"), np.datetime64("NaT", "ns")), we are actually passing a new dt64 object.  When doing the dict lookup it correctly gets hash-equality and then checks for list-equality, which then shows up as False.  So we get cache misses.

Let's check this out...  At the module-level in dtypes.cast I define `dt64nat = np.datetime64("NaT")` and in maybe_promote I do

```
    if isinstance(fill_value, np.datetime64) and isna(fill_value):
        fill_value = dt64nat
```

Then using the example from https://github.com/pandas-dev/pandas/pull/39692#issuecomment-1439203636

```
In [2]: %timeit right = df[mask]
182 µs ± 7.12 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

about 8 microseconds slower than the disabling-the-cache measurement, which seems reasonable for the extra check added.  Calling this a win.

Presumably we'd see the same issue with td64nat or if any case we passed a NaN that wasn't specifically `np.nan` (which is probably rare).
